synapse/storage/databases/main/events.py

```diff
@@ -1943,15 +1943,29 @@ def _handle_batch_event(self, txn: LoggingTransaction, event: EventBase):
 
         txn.execute(sql, (batch_id,))
 
-    def _handle_redaction(self, txn, redacted_event_id):
+    def _handle_redact_relations(
+        self, txn: LoggingTransaction, redacted_event_id: str
+    ) -> None:
         """Handles receiving a redaction and checking whether we need to remove
         any redacted relations from the database.
 
         Args:
             txn
-            redacted_event_id (str): The event that was redacted.
+            redacted_event_id: The event that was redacted.
         """
 
+        # Fetch current relations.
+        relates_to = self.db_pool.simple_select_one_onecol_txn(
+            txn,
+            table="event_relations",
+            keyvalues={"event_id": redacted_event_id},
+            retcol="relates_to_id",
+            allow_none=True,
+        )
+        # Any relation information for the related events must be cleared.
+        if relates_to is not None:
+            txn.call_after(self.store._invalidate_relations_for_event, relates_to)
```

Comment:
This will only redact this properly on the process which is persisting events (I believe) -- any ideas on how to stream this properly to other servers?

Reply:
There are two options:

1. Call `self.store._invalidate_cache_and_stream` for each cache you want to invalidate (instead of `txn.call_after(cache.invalidate)`. This both invalidates the cache locally and streams the invalidation to other workers. Or
2. Do what we do for current state event invalidation and send a special `RELATION_REDACTED` cache invalidation over replication, which then other workers will see and invalidate all the right caches:

https://github.com/matrix-org/synapse/blob/a8da0469070771b0b63c97a96f3221afea2aa2e7/synapse/storage/databases/main/cache.py#L42-L44

The latter method is really only worth it if we're frequently invalidating lots of caches at once

Reply:
I suspect the former is the way to go then, I don't see this as being a super frequent?

Reply:
Done in 6687bbd8b524cb9f6f7751f2b0509a3d636c6a50, thanks!